### ISSUE TITLE
The performance of toBuffer will drop after overlayWith a small png file.

### ISSUE THREAD

**Issue by hongyuanlei** [lovell/sharp#1027]
I did a test about the `toBuffer` method of sharp.

First, I test the `toBuffer` without `overlayWith`. The test result is:

2017-11-16T04:42:32.026Z - info: ended_to_buffer_image:  383
2017-11-16T04:42:32.037Z - info: ended_to_buffer_image:  392
2017-11-16T04:42:32.071Z - info: ended_to_buffer_image:  117
2017-11-16T04:42:32.105Z - info: ended_to_buffer_image:  150
2017-11-16T04:42:32.116Z - info: ended_to_buffer_image:  161
2017-11-16T04:42:32.129Z - info: ended_to_buffer_image:  142
2017-11-16T04:42:32.153Z - info: ended_to_buffer_image:  166
2017-11-16T04:42:32.167Z - info: ended_to_buffer_image:  179
2017-11-16T04:42:32.184Z - info: ended_to_buffer_image:  192
2017-11-16T04:42:32.222Z - info: ended_to_buffer_image:  228
2017-11-16T04:42:32.225Z - info: ended_to_buffer_image:  235
2017-11-16T04:42:32.248Z - info: ended_to_buffer_image:  252
2017-11-16T04:42:32.254Z - info: ended_to_buffer_image:  257

Then, I test the `toBuffer` with `overlayWith` another image(It's a png image, size is 3k). I cached the overlay image buffer. Each time of:
```
rawImage.overlayWith(buffer, {
              tile: filterRequest.overlayTile(),
              gravity: gravity,
              top: position.top,
              left: position.left
            });
```
it will use the same overlay image buffer. I test the time used of `overlayWith` is 0 or 1ms.

Here's the time used of `toBuffer` method:

 2017-11-15T08:17:47.862Z - info: ended_to_buffer_image:  564
 2017-11-15T08:17:47.928Z - info: ended_to_buffer_image:  741
 2017-11-15T08:17:48.083Z - info: ended_to_buffer_image:  764
 2017-11-15T08:17:48.182Z - info: ended_to_buffer_image:  518
 2017-11-15T08:17:48.344Z - info: ended_to_buffer_image:  675
 2017-11-15T08:17:48.415Z - info: ended_to_buffer_image:  744
 2017-11-15T08:17:48.543Z - info: ended_to_buffer_image:  850
 2017-11-15T08:17:48.709Z - info: ended_to_buffer_image:  1015
 2017-11-15T08:17:48.821Z - info: ended_to_buffer_image:  1126
 2017-11-15T08:17:48.860Z - info: ended_to_buffer_image:  1163
 2017-11-15T08:17:49.025Z - info: ended_to_buffer_image:  1326
 2017-11-15T08:17:49.175Z - info: ended_to_buffer_image:  1472
 2017-11-15T08:17:49.263Z - info: ended_to_buffer_image:  1556
 2017-11-15T08:17:49.302Z - info: ended_to_buffer_image:  1589
 2017-11-15T08:17:49.517Z - info: ended_to_buffer_image:  1796
 2017-11-15T08:17:49.635Z - info: ended_to_buffer_image:  1910
 2017-11-15T08:17:49.703Z - info: ended_to_buffer_image:  1975
 2017-11-15T08:17:49.830Z - info: ended_to_buffer_image:  2094
 2017-11-15T08:17:49.948Z - info: ended_to_buffer_image:  2207
 2017-11-15T08:17:50.091Z - info: ended_to_buffer_image:  2343 

As you can see, the performance of `toBuffer` drop. 

VIPS version 8.5.5. 
Node version 8.9.1. 
Sharp version 0.18.4.

**Comment by lovell** [comment]
> "I test the time used of overlayWith is 0 or 1ms."

This function sets the values to be used later when `toBuffer` is called so this is expected.

The timings suggest you may be initiating all the `toBuffer` calls at once rather than (a)waiting for each one to return before the next call to `toBuffer`.

Did you see http://sharp.dimens.io/en/stable/api-utility/#concurrency ? By default 4 images are processed in parallel.

As an aside the `overlayWith` code will be undergoing a fairly major refactor as part of #728 and will provide the repeated overlay scenario you're testing.

**Comment by hongyuanlei** [comment]
@lovell Thanks for your reply.

I use ec2 c4.large to run the test. So the default concurrency count should be 2.

What confuses me is that the size of raw image info will increase 6 times after use `overlayWith`. I guess this makes the `overlayWith` slow.

**Comment by hongyuanlei** [comment]
HI @lovell, Do you know why the size of raw image info will increase maybe 6 times after use `overlayWith`?

**Comment by lovell** [comment]
Please can you provide a more complete, standalone code sample with the time measurements in place for me to be able to help further.

**Comment by hongyuanlei** [comment]
@lovell Thank you. I guess I know why `overlayWith` become slow. Becuase the overlay image that I use is a png file, it contains an alpha channel. 

This comes from the document:
`If the overlay image contains an alpha channel then composition with premultiplication will occur`.

If I use a jpg overlay image, then `toBuffer` will be very quick. But my problem is I need the overlay image has transparent space. I don't know if there are some better ways to implement it.

**Comment by lovell** [comment]
Please subscribe to #728 for updates about improvements to `overlayWith`, which I think will help you, but without a more complete code sample it's quite hard to tell what you're trying to achieve.

**Comment by hongyuanlei** [comment]
Hi @lovell ,  My scenario is that I have a web application it depends on "sharp" and it can resize and overlay a watermark of the requested image then return to the client side. There are lots of image request from the client, So I need the performance of this web application to be very high. Previously, We didn't have the requirement of the watermark and our web application just has the resize functionality, the performance is very high. For now, we have the requirement to dynamically add a watermark. The problem is that the watermark is png image and it contains an alpha channel, so it will cause:

`If the overlay image contains an alpha channel then composition with premultiplication will occur.`

Watermarking reduces performance maybe 6x after overlay with a png watermark. I know this is also related to the complexity of watermarking pictures.  My question is if there are some better ways to overlay with a watermark image which contains an alpha channel?

The code is like this:

```
const processImage = (rawImage, cb) => {
  rawImage
    .resize(
      400,
      400,
      {interpolator: sharp.interpolator.bicubic}
    );

  sharp('/overlay/logo.png').toBuffer((err, overlayImage, metadata) => {
    if (err) {
      return cb(err);
    }

    rawImage.overlayWith(buffer, {
      gravity: 'southeast',
      top: xxx,
      left: xxx
    })
    .toFormat('jpeg');

    return cb(null, rawImage);
  });	
}
 
... 
processImage(img, (err, newImg) => {
  if (err) {
    return cb(err);
  }

  const start_to_buffer_image = +new Date();
  newImg.toBuffer((err, output, infoResized) => {
    if (err) {
      return cb(err);
    }

    const ended_to_buffer_image = +new Date();
    logger.log.info("ended_to_buffer_image: ", (ended_to_buffer_image - start_to_buffer_image));

    cb(err, output);
  });

}); 
...

```

**Comment by hongyuanlei** [comment]
HI @lovell , Have you done a performance test of overlay an image which contains an alpha channel before?

**Comment by lovell** [comment]
Thank you for providing more complete example code..

Try providing the `'overlay/logo.png'` path directly to `overlayWith` rather than via
`sharp('/overlay/logo.png').toBuffer(...` as you're incurring a costly and unnecessary PNG decode/encode roundtrip.
```diff
-sharp('/overlay/logo.png').toBuffer((err, overlayImage, metadata) => {
-    if (err) {
-      return cb(err);
-    }
-
-    rawImage.overlayWith(overlayImage, {
+    rawImage.overlayWith('/overlay/logo.png', {
```

**Comment by hongyuanlei** [comment]
@lovell Thank you, but after use path directly the time used of `toBuffer` is the same:

```
  const start_to_buffer_image = +new Date();
  newImg.toBuffer((err, output, infoResized) => {
    if (err) {
      return cb(err);
    }

    const ended_to_buffer_image = +new Date();
    logger.log.info("ended_to_buffer_image: ", (ended_to_buffer_image - start_to_buffer_image));

    cb(err, output);
  });

```

Without overlay `toBuffer` just need nearly 15ms, but with `overlayWith` png file then `toBuffer`, the time used will go to 90ms or more.

**Comment by lovell** [comment]
I'd expect `overlayWith` to involve more CPU time as it's doing more work.

This module uses `uv_queue_work` to access a shared worker thread pool. If your code is processing more than 4 images at once then the timings captured by this sample code will also include the time queued waiting for a libuv thread to service the `toBuffer` call.

If you've not already done so, please read http://sharp.dimens.io/en/stable/api-utility/#concurrency

See also this Stack Overflow answer https://stackoverflow.com/a/29472222/4625124

**Comment by lovell** [comment]
I hope this answered your questions, please re-open with more details if not.
